### PR TITLE
[solution]:zh/medium-return-type

### PR DESCRIPTION
**Pull request by SUtengda**
# Challenge [CHALLENGE_NAME]

## Checklist

- [x] Make sure that you are opening the PR with a single solution (one solution per PR)
- [x] The solution must have at least several references for the readers to dive in depth

### CODE REVIEW

File: zh/medium-return-type.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  if (v) return 1;
+  else return 2;
+};
+
+type a = MyReturnType<typeof fn>; // should be "1 | 2"
+```
+
+## 解答
+
+使用[条件类型中的类型推导](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types)的经验法则是当你不确定类型必须是什么的时候。
```
Comment (likui628) [inline zh/medium-return-type.md:26]: 在[条件类型中使用类型推导](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types)的经验法则是当你不确定类型必须是什么的时候。

File: zh/medium-return-type.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+```ts
+type MyReturnType<T> = T extends () => infer R ? R : T;
+```
+
+如果我们的类型 `T` 可以分配给函数，我们推断它的返回类型并将其返回，否则我们返回 `T` 本身。
+比较直截了当。
+
+这个解决方案的问题是，如果我们传递一个带参数的函数，它将不能分配给我们的类型 `() => infer R`。
+
+让我们通过添加 `...args: any[]` 来证明我们可以接受任何参数并且我们不关心它们：
```
Comment (likui628) [inline zh/medium-return-type.md:43]: 证明->表明

Comment (SUtengda) [reply zh/medium-return-type.md]: done, thanks

File: zh/medium-return-type.md (hunk 1 above)
Comment (SUtengda) [reply zh/medium-return-type.md]: done

**Review by likui628 — APPROVED** [review]

**Comment by charleserious** [comment]
> # Challenge [CHALLENGE_NAME]
> ## Checklist
> * [x]  Make sure that you are opening the PR with a single solution (one solution per PR)
> * [x]  The solution must have at least several references for the readers to dive in depth

@SUtengda don't forget to change the title `Challenge [CHALLENGE_NAME]` into things like `Challenge [zh/medium-return-type.md]` ☕

**Pull request merged**